pkg/kubelet/kubelet.go

```diff
@@ -2845,6 +2855,117 @@ func (kl *Kubelet) fastStatusUpdateOnce() {
 	}, 100*time.Millisecond, stopCh)
 }
 
+// handleMaxContainerCheckpoints goes through the list of checkpoints of a
+// container in a pod and deletes old checkpoints if there are more than
+// 'kl.maxContainerCheckpoints'.
+func (kl *Kubelet) handleMaxContainerCheckpoints(
+	podUID types.UID,
+	podFullName,
+	containerName string,
+	options *runtimeapi.CheckpointContainerRequest,
+) error {
+
+	if kl.maxContainerCheckpoints <= 0 {
+		// This cannot really happen as the kubelet config parser will
+		// error out early if this is <= 0. This additional error
+		// handling is mainly here to make sure tests are working.
+		return errors.New("maxContainerCheckpoints must be greater than 0")
+	}
+
+	// Lock while accessing the checkpoints on disk to avoid chaos
+	// if multiple threads would do it at the same time.
+	kl.maxContainerCheckpointsLock.Lock()
+	defer kl.maxContainerCheckpointsLock.Unlock()
+
+	checkpointDirectory, err := os.ReadDir(kl.getCheckpointsDir())
+	if err != nil {
+		return fmt.Errorf(
+			"error reading checkpoint archive directory %s: %w",
+			kl.getCheckpointsDir(),
+			err,
+		)
+	}
+
+	regexString := fmt.Sprintf(
+		`checkpoint-%s-%s-\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.*tar`,
+		podFullName,
+		containerName,
+	)
+
+	re, err := regexp.Compile(regexString)
+	if err != nil {
+		return fmt.Errorf(
+			"error compiling regex %s: %w",
+			regexString,
+			err,
+		)
+	}
+
+	var checkpointArchives = []string{}
+	for _, archive := range checkpointDirectory {
+		// only select checkpoint archives which have been created from the same
+		// namespace/pod/container combination
+		if re.MatchString(archive.Name()) {
+			checkpointArchives = append(
+				checkpointArchives,
+				filepath.Join(
+					kl.getCheckpointsDir(),
+					archive.Name()),
```

Comment:
I'd add `archive` instead of a path. It already has a `ModTime`, so we can sort it without calling Stat.

Reply:
`archive.Info()` used as suggested